resources/execute_event_handler.rb

```diff
@@ -0,0 +1,182 @@
+# frozen_string_literal: true
+
+resource_name :execute_event_handler
+provides :execute_event_handler
+
+# Resource to execute BYOS event handler
+
+property :event_name, String, name_property: true
+property :event_command, String, required: false
+
+default_action :run
+
+action :run do
+  if new_resource.event_command.nil?
+    Chef::Log.info("No command defined for Event #{new_resource.event_name}, noop")
+    return
+  end
+
+  event_log = '/var/log/parallelcluster/byos-plugin.log'
+  event_cwd = '/home/byos'
+  event_user = 'byos'
+  event_env = build_env
+  event_log_prefix_error = "[%Y-%m-%d %H:%M:%S] ERROR [#{new_resource.event_name}]"
+  event_log_prefix_info = "[%Y-%m-%d %H:%M:%S] INFO [#{new_resource.event_name}]"
+  Chef::Log.info("Executing Event #{new_resource.event_name}, with user (#{event_user}), cwd (#{event_cwd}), command (#{new_resource.event_command}), log (#{event_log})")
+  # shellout https://github.com/chef/mixlib-shellout
+  # switch stderr/stdout with (2>&1 1>&3-), process error (now on stdout), switch back stdout/stderr with (3>&1 1>&2) and then process output
+  cmd = Mixlib::ShellOut.new("set -o pipefail; { (bash #{new_resource.event_command}) 2>&1 1>&3- | ts '#{event_log_prefix_error}' | tee -a #{event_log}; } " \
+    "3>&1 1>&2 | ts '#{event_log_prefix_info}' | tee -a #{event_log}", user: event_user, env: event_env, cwd: event_cwd)
+  cmd.run_command
+
+  if cmd.error?
+    raise "Expected Event #{new_resource.event_name} to exit with #{cmd.valid_exit_codes.inspect}," \
+      " but received '#{cmd.exitstatus}', complete log in #{event_log}\n #{format_stderr(cmd)}"
+  end
+end
+
+action_class do # rubocop:disable Metrics/BlockLength
+  def format_stderr(cmd)
+    "---- STDERR for #{new_resource.event_name} Event ----\n" \
+    "#{cmd.stderr.strip}\n" \
+    "---- End STDERR for #{new_resource.event_name} Event ----\n"
+  end
+
+  def build_env
+    # TODO: move create dir where setting up user/byos/scheduler
+    FileUtils.mkdir_p('/home/byos/.parallelcluster/')
+
+    source_cluster_config = node.dig(:cluster, :cluster_config_path)
+    target_cluster_config = '/home/byos/.parallelcluster/cluster-config.yaml'
+    if source_cluster_config && !source_cluster_config.empty?
+      FileUtils.cp(source_cluster_config, target_cluster_config)
+    end
```

Comment:
I'm wondering if we'd better fail when the source configs are not found

Reply:
done